[export] add runtime assert messages to python torch checks

pianpwk opened the pull request:
~fixes #150063 (for python at least)

Before:
```
Runtime assertion failed for expression Eq(Mod(s16*s35, s35 - 1), 0) on node 'eq'
```

Now:
```
RuntimeError: Runtime assertion failed for expression Eq(Mod(s16*s35, s35 - 1), 0) on node 'eq'

The original traceback points to the following location and error message:
/data/users/pianpwk/pytorch/torch/_prims_common/__init__.py:954
shape '[s35 - 1, ((s16*s35)//(s35 - 1))]' is invalid for input of size s16*s35
```

Differential Revision: D72483950

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

laithsakka commented on torch/__init__.py:
```diff
@@ -1629,7 +1629,17 @@ def _check_with(
+                pass
+        else:
+            msg = message
+
```
yeh maybe  we materialize it at the point were we insert a deferred_runtime_asserts_over_guards

laithsakka commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -258,14 +258,17 @@ def add_runtime_asserts(ras):
                 with _set_node_metadata_hook(gm, _node_metadata_hook):
                     res = _sympy_interp(expr_to_proxy, ra.expr).node
+                    expr_msg = f"Runtime assertion failed for expression {ra.expr} on node '{res}'"
+                    assert_msg = (
```
we need something like this in inductor  also for torch.compile. maybe put this inside a function in ra to avoid duplication?
see the change in ir.py in this file

laithsakka commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -258,14 +258,17 @@ def add_runtime_asserts(ras):
+                    )
                     graph.call_function(
                         torch.ops.aten._assert_scalar.default,
                         # TODO: use ra.msg here, but it's pretty
```
do we need this to do to still be here?

laithsakka commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -258,14 +258,24 @@ def add_runtime_asserts(ras):
+                            assert isinstance(ra.msg, str)
+                            ra_msg = ra.msg
+                    if ra_msg is not None:
+                        assert_msg = f"{expr_msg}\n\nThe original traceback points to the following location and error message:\n{ra.loc}\n{ra_msg}"
```
maybe with out new lines? this is how it show up when the graph is printed 
![Uploading Screenshot 2025-04-22 at 8.36.47 AM.png…]()

laithsakka commented on test/export/test_export.py:
```diff
@@ -13117,6 +13119,13 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            r"torch\/_prims_common\/__init__.py:(.*\n)*.*" + \
+            r"shape .* is invalid for input of size .*",
+        ):
+            ep.module()(torch.randn(9, 9))
```
i thought with r" you do not need \ to escape (?

laithsakka commented:
@pytorchbot merge

pianpwk committed:
[export] add runtime assert messages to python torch checks (#150719)

Summary:
~fixes https://github.com/pytorch/pytorch/issues/150063 (for python at least)

Before:
```
Runtime assertion failed for expression Eq(Mod(s16*s35, s35 - 1), 0) on node 'eq'
```

Now:
```
RuntimeError: Runtime assertion failed for expression Eq(Mod(s16*s35, s35 - 1), 0) on node 'eq'

The original traceback points to the following location and error message:
/data/users/pianpwk/pytorch/torch/_prims_common/__init__.py:954
shape '[s35 - 1, ((s16*s35)//(s35 - 1))]' is invalid for input of size s16*s35
```


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv


Reviewed By: laithsakka

Differential Revision: D72483950

Pulled By: pianpwk
